### PR TITLE
ARROW-9188: [C++] Use Brotli shared libraries if they are available

### PR DESCRIPTION
**Pull request by wesm**
If both shared and static Brotli libraries are available, the static ones were being selected, causing ~750KB of code to be statically linked into libarrow.so on Linux. This is not consistent with our handling of other toolchain libraries. We should use the shared library if it is available.

### CODE REVIEW

File: cpp/cmake_modules/FindBrotli.cmake [hunk 1]
```diff
@@ -17,29 +17,29 @@
 #
 #  find_package(Brotli)
 
-# Favour static libraries over dynamic libraries, and handle various spellings
+# Favor shared libraries over dynamic libraries, and handle various spellings
```
Comment (kiszk) [inline cpp/cmake_modules/FindBrotli.cmake:20]: nit: dynamic -> static ?

**Comment by kiszk** [comment]
Looks good except one minor comment. LZ4 and ZSTD also use the dynamic library at first if available.

**Comment by wesm** [comment]
Thanks, will look into this. I'm guessing these changes will break some of the Python wheel builds so we may need a flag to indicate a preference of shared vs static

**Comment by wesm** [comment]
@github-actions crossbow submit -g linux -g wheel -g conda

**Comment by wesm** [comment]
It appears that the Brotli shared libraries are in the manylinux1 image even though `-DBUILD_SHARED_LIBS=OFF` 

https://github.com/apache/arrow/blob/master/python/manylinux1/scripts/build_brotli.sh#L29

**Comment by wesm** [comment]
@nealrichardson I figure this might impact the R packages also

**Comment by nealrichardson** [comment]
@wesm how can I help/what should I look for?

**Comment by wesm** [comment]
If you're installing Brotli in any of the packaging setups, there may be a scenario where there is both the shared AND static library -- in that case there would be an issue. We need to see why the manylinux* builds have `libbrotli*.so` in them -- if for some reason Brotli's build is insubordinate and churlish we might have to add a flag to do a hard switch between shared/static like we do with ZSTD

**Comment by nealrichardson** [comment]
IIUC we're ok:

* Windows: no brotli: https://github.com/apache/arrow/blob/master/ci/scripts/PKGBUILD
* macOS: no brotli: https://github.com/apache/arrow/blob/master/dev/tasks/homebrew-formulae/autobrew/apache-arrow.rb
* Linux: maybe brotli, but it's all static build anyway: https://github.com/apache/arrow/blob/master/r/inst/build_arrow_static.sh

**Comment by wesm** [comment]
Apparently the `-DBUILD_SHARED_LIBS=OFF` option for Brotli doesn't do anything. I'll add some code to scrub the shared libs from the manylinux images

**Comment by wesm** [comment]
Actually, that's crazy. I'm taking the same approach as ZSTD and adding a CMake toggle between shared and static Brotli (with default being shared)

**Comment by wesm** [comment]
@github-actions crossbow submit -g linux -g wheel -g conda

**Comment by wesm** [comment]
This looks good to me, @kszucs @kou @nealrichardson anything else you would want to check?

**Comment by wesm** [comment]
If there's no further feedback I will merge this in the next 24h and I assume that any packaging issues will come up in nightlies as we push toward the next release.

File: cpp/cmake_modules/DefineOptions.cmake [hunk 2]
```diff
@@ -285,6 +285,9 @@ if("${CMAKE_SOURCE_DIR}" STREQUAL "${CMAKE_CURRENT_SOURCE_DIR}")
   define_option(ARROW_PROTOBUF_USE_SHARED
                 "Rely on Protocol Buffers shared libraries where relevant" ON)
 
+  define_option(ARROW_BROTLI_USE_SHARED "Rely on Brotli shared libraries where relevant"
+                ON)
+
```
Comment (kou) [inline cpp/cmake_modules/DefineOptions.cmake:290]: Could you keep `ARROW_*_USE_SHARED` in alphabetical order?

File: python/manylinux1/build_arrow.sh [hunk 3]
```diff
@@ -122,6 +122,7 @@ cmake \
     -DARROW_WITH_SNAPPY=ON \
     -DARROW_WITH_ZLIB=ON \
     -DARROW_WITH_ZSTD=ON \
+    -DARROW_BROTLI_USE_SHARED=OFF \
```
Comment (kou) [inline python/manylinux1/build_arrow.sh:125]: Could you keep this option list in alphabetical order?

File: python/manylinux201x/build_arrow.sh [hunk 4]
```diff
@@ -120,6 +120,7 @@ PATH="${CPYTHON_PATH}/bin:${PATH}" cmake \
     -DARROW_WITH_SNAPPY=ON \
     -DARROW_WITH_ZLIB=ON \
     -DARROW_WITH_ZSTD=ON \
+    -DARROW_BROTLI_USE_SHARED=OFF \
```
Comment (kou) [inline python/manylinux201x/build_arrow.sh:123]: Ditto.

**Review by kou — APPROVED** [review]
+1

**Pull request closed without merging**